mantidproject/mantid
# Transverse filed (TF) fit functions

**Description of work**
The Transverse field (TF) fit function is used to calculate the normalisation of muon asymmetry data. For this PR I have created an algorithm that takes the user function f (e.g. a sinusoidal function) and wrap it for the normalisation:

N(1+f)+expMuonDecay

By default we set the amplitude of ExpMuonDecay to zero. N is the normalisation. 

The extract mode will take the above function and return the user function f. 
 
**To test:**
There are a few scripts below that depict the use of the algorithm. 

1D example, the fixes and ties should at the end and in `fittedFunction `
```
#create a normalisation table
tab = CreateEmptyTableWorkspace()
tab.addColumn('double', 'norm')
tab.addColumn('str', 'name')
tab.addColumn('str', 'method')

tab.addRow([1.,"a","Estimate"])
tab.addRow([2.,"Run;;Group;;b;;Asym;;#1","Estimate"])
tab.addRow([3.,"Run;;Group;;c;;Asym;;#1","Estimate"])
tab.addRow([4.,"Run;;Group;;d;;Asym;;#1","Estimate"])

#create original function and workspace
myFunc='name=LinearBackground,A0=3,A1=4;name=LinearBackground,A0=0,A1=2;ties=(f0.A1=3, f0.A0=f1.A0)'
ws = CreateWorkspace(DataX=[1,2,3,4,5,6,7], DataY=[1,2,3,4,5,6],OutputWorkspace="a")

TFFunc = ConvertFitFunctionForMuonTFAsymmetry(myFunc,tab,["a"],"Construct")

# do a fit with new function

fit =mantid.AlgorithmManager.create("Fit")
fit.setProperty("Function",str(TFFunc))
fit.setProperty("InputWorkspace",ws)
fit.setProperty("Output",'fitWS')
fit.execute()
fittedFunc = fit.getPropertyValue("Function")

returnFunc = ConvertFitFunctionForMuonTFAsymmetry(str(fittedFunc),tab,["a"],"Extract")
# 0 iteration fit to get param table -> wont change function values
fit_output = Fit(Function=str(returnFunc),InputWorkspace=ws,MaxIterations=0,Output="return_params")

paramTable = fit_output.OutputParameters  # table containing the optimal fit parameters
print('Gradients (shared):')
print("Gradient 1: {0:.2f}".format(paramTable.column(1)[3]))
print("Gradient 2: {0:.2f}".format(paramTable.column(1)[1]))
print('offsets:')
print("Constant 1: {0:.2f}".format(paramTable.column(1)[0]))
print("Constant 2: {0:.2f}".format(paramTable.column(1)[2]))
```

Multi fitting example:

```
import math
import numpy as np

# create data
xData=np.linspace(start=0,stop=10,num=22)
yData=[]
for x in xData:
    yData.append(2.0*x+10.)
yData2=[]
for x in xData:
    yData2.append(5.0*x+7.)
# create workspaces
input = CreateWorkspace(xData,yData,OutputWorkspace="Run; Group; a; Asym; #1")
input2 = CreateWorkspace(xData,yData2,OutputWorkspace="Run; Group; b; Asym; #1")
# create function
myFunc=';name=LinearBackground,$domains=i,A0=0,A1=1'
myFunc2=';name=LinearBackground,$domains=i,A0=2,A1=3'
multiFunc='composite=MultiDomainFunction,NumDeriv=1'+myFunc+myFunc2 +';ties=(f0.A1=f1.A1,f0.A0=1.23)'


tab = CreateEmptyTableWorkspace()
tab.addColumn('double', 'norm')
tab.addColumn('str', 'name')
tab.addColumn('str', 'method')

tab.addRow([1.,"Run;;Group;;a;;Asym;;#1","Estimate"])
tab.addRow([2.,"Run;;Group;;b;;Asym;;#1","Estimate"])
tab.addRow([3.,"Run;;Group;;c;;Asym;;#1","Estimate"])
tab.addRow([4.,"Run;;Group;;d;;Asym;;#1","Estimate"])



bob= ConvertFitFunctionForMuonTFAsymmetry(multiFunc,tab,["Run; Group; a; Asym; #1","Run; Group; b; Asym; #1","Run; Group; b; Asym; #1"],"Construct")

print "boib"
print bob
print
out = Fit(Function=multiFunc, InputWorkspace=input, WorkspaceIndex=0, \
                 InputWorkspace_1=input2, WorkspaceIndex_1=0, \
                 StartX = 0.1, EndX=9.5, StartX_1 = 0.1, EndX_1=9.5,Output='fit_original')


fit =mantid.AlgorithmManager.create("Fit")
fit.setProperty("Function",str(bob))
fit.setProperty("InputWorkspace",input)
fit.setProperty("WorkspaceIndex",0)
fit.setProperty("InputWorkspace_1",input2)
fit.setProperty("WorkspaceIndex_1",0)
fit.setProperty("StartX" ,0.1)
fit.setProperty("EndX",9.5)
fit.setProperty("StartX_1" ,0.1)
fit.setProperty("EndX_1",9.5)
fit.setProperty("Output",'fit2')
fit.execute()
a = fit.getPropertyValue("Function")

print a
b=ConvertFitFunctionForMuonTFAsymmetry(a,tab,["Run; Group; a; Asym; #1","Run; Group; b; Asym; #1"],"Extract")
print
print b
fit_output = Fit(Function=b, InputWorkspace=input, WorkspaceIndex=0, \
                 InputWorkspace_1=input2, WorkspaceIndex_1=0, \
                 StartX = 0.1, EndX=9.5, StartX_1 = 0.1, EndX_1=9.5,Output='fit')
paramTable = fit_output.OutputParameters  # table containing the optimal fit parameters
# print results
print('Gradients (shared):')
print("Gradient 1: {0:.2f}".format(paramTable.column(1)[3]))
print("Gradient 2: {0:.2f}".format(paramTable.column(1)[1]))
print('offsets:')
print("Constant 1: {0:.2f}".format(paramTable.column(1)[0]))
print("Constant 2: {0:.2f}".format(paramTable.column(1)[2]))
```



Fixes #22423. 

**Release Notes** 

In release notes
---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
